### PR TITLE
Fix #323: Added additional activation OTP documentation.

### LINKED ISSUE #323: Additional activation OTP
The purpose of additional activation OTP is to help with the user authentication, or with the activation confirmation. The additional OTP can be used either in the early stages of the activation or later when the activation is created and waits for the confirmation. The proposed change has also impact on the recovery activations and allows an additional confirmation of such type of activation

We'll use this ticket as reference information about the feature. It's expected that this feature has no direct impact on powerauth-crypto code. So, only the documentation will be updated.

# Feature description

## Additional user authentication

In this scenario, it's expected that the PowerAuth activation is not yet created so that the additional activation OTP can be used in the time of the activation creation. The typical activation flow would look like this

1. User is properly authenticated in the internet banking and initiates the activation creation.
   1. IB generates random activation OTP
   1. IB then asks PowerAuth server to create an activation, with using additional OTP (`initActivation`)
   1. IB displays QR code, containing an activation code
   1. At the same time, IB initiates the activation OTP delivery, for example via SMS.
1. In the mobile application 
   1. User scans QR code or retypes the activation code manually.
   1. User waits for OTP code delivery via out-of-band channel.
   1. User retypes the OTP code to the mobile application.
1. Mobile application then initializes the activation, using activation code and OTP. The activation code and OTP is verified in `prepareActivation`.

## Activation confirmation

In this scenario, the OTP helps with the final activation confirmation, so the OTP is required in the later stages of the activation process (during commit). Then the typical activation flow may look like:

1. User is properly authenticated in the internet banking and initiates the activation creation.
   1. IB then asks PowerAuth server to create an activation, with no additional OTP
   1. IB displays a QR code, containing an activation code
2. In the mobile application 
   1. User scans QR code or retypes the activation code manually.
3. Mobile application then initializes the activation, with using activation code.
4. After the activation keys are exchanged (`prepareActivation` or `createActivation`)
   1. IB generates activation OTP and initiates its delivery, for example via SMS.
   1. IB uses `updateActivationOtp` to associate OTP with the pending activation. 

Now it depends whether the OTP must be retyped back to the internet banking, or the mobile application. For the first case, the implementation is straightforward. Once the user retypes OTP back to IB, the activation can be completed with `commitActivation`. In case the commit fails, we can allow the user to retype OTP, or go back to step 4. For both cases, the number of commit attempts is limited to the predefined maximum number of failed attempts.

In case the OTP is retyped in the mobile application, the additional RESTful endpoint has to be implemented on the server that provides data channels for the mobile app. In case that commit fails, then we can allow the user to retype OTP, or go back to step 4. For both, the number of commit attempts is limited to the predefined maximum number of failed attempts.

A very similar scenario can also be implemented for the recovery code activations.

### PR DESCRIPTION
**Pull request by hvge**

### CODE REVIEW

File: docs/Additional-Activation-OTP.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+   1. The user waits for OTP delivery via the out-of-band channel.
+   1. The user retypes OTP.
+   1. Mobile application then commits the activation with OTP, by calling a custom RESTful endpoint, protected with our [ECIES encryption](End-To-End-Encryption.md) scheme.
+
+1. Intermediate Server Application then receives the commit request
+
+   1. Decrypts OTP from the request
+   1. Commits the activation by calling PowerAuth Server's [`commitActivation`](https://github.com/wultra/powerauth-server/blob/develop/docs/SOAP-Service-Methods.md#method-commitactivation) method.
+
+After the response from the commit is received on the mobile side, the application can check the  whether the activation's state is ACTIVE.
```
Comment (romanstrobl) [inline docs/Additional-Activation-OTP.md:168]: Typo: `can check the  whether` (the extra space typo is not visible)

File: docs/Additional-Activation-OTP.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+1. Intermediate Server Application receives a recovery activation request:
+
+   1. The login and password is verified by the PowerAuth Server, by calling [`recoveryCodeActivation`](https://github.com/wultra/powerauth-server/blob/develop/docs/SOAP-Service-Methods.md#method-recoverycodeactivation).
+   1. Intermediate Server Application generates random activation OTP and update the activation record, by calling [`updateActivationOtp`](https://github.com/wultra/powerauth-server/blob/develop/docs/SOAP-Service-Methods.md#method-updateactivationotp) method.
+   1. At the same time, Intermediate Server Application initiates the delivery of activation OTP.
+   
+1. Back in the mobile application:
+
+   1. The mobile application receives the response from the server and completes the keys-exchange on the mobile side.
```
Comment (romanstrobl) [inline docs/Additional-Activation-OTP.md:158]: The `keys-exchange` should probably be rather `key exchange`, this is a term in cryptography.

File: docs/Additional-Activation-OTP.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+After the response from the commit is received on the mobile side, the application can check the  whether the activation's state is ACTIVE.
+
+### Confirm activation recovery with OTP
+
+The confirmation of activation recovery is very similar to custom activation confirmation.
+
+1. In the mobile application:
+   
+   1. The user enters recopvery code and PUK
```
Comment (romanstrobl) [inline docs/Additional-Activation-OTP.md:147]: Typo: `recopvery`

File: docs/Additional-Activation-OTP.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+   1. The user waits for OTP delivery via the out-of-band channel.
+   1. The user retypes OTP.
+   1. Mobile application then commits the activation with OTP, by calling a custom RESTful endpoint, protected with our [ECIES encryption](End-To-End-Encryption.md) scheme.
+
+1. Intermediate Server Application then receives the commit request
+
+   1. Decrypts OTP from the request
+   1. Commits the activation by calling PowerAuth Server's [`commitActivation`](https://github.com/wultra/powerauth-server/blob/develop/docs/SOAP-Service-Methods.md#method-commitactivation) method.
+
+After the response from the commit is received on the mobile side, the application can check the  whether the activation's state is ACTIVE.
```
Comment (romanstrobl) [inline docs/Additional-Activation-OTP.md:139]: Typo: `check the  whether` (extra space typo not visible)

File: docs/Additional-Activation-OTP.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+You can also slightly alter this whole sequence, and generate the first OTP later, in the step 3. In this case, you have to use [`updateActivationOtp`](https://github.com/wultra/powerauth-server/blob/develop/docs/SOAP-Service-Methods.md#method-updateactivationotp) in the same step, to set the OTP. 
+
+### Confirm custom activation with OTP
+
+There are multiple ways how to implement custom activation and confirm it with an additional authentication OTP so that we will discuss only one particular case. In this scenario, we don't involve the Master Front-End Application in the process, but we expect that the user has already issued some valid credentials to the system. 
+
+1. In the mobile application:
+   
+   1. The user enters login and password
```
Comment (romanstrobl) [inline docs/Additional-Activation-OTP.md:117]: Not sure about the `login` word, it is better to use `username` or similar term.

File: docs/Additional-Activation-OTP.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+1. Let's say that the user is authenticated in Master Front-End Application (with login and password) and initiates the activation creation process:
+
+   1. Master Front-End Application generates random activation OTP and keeps it temporarily in the database.
+   1. At the same time, Master Front-End Application initiates the delivery of activation OTP. It's recommended to deliver such code via the dedicated out-of-band channel, for example, via SMS.
+   1. Master Front-End Application instructs the user to start the mobile application and type the login, password, and OTP to the mobile app.
+   
+1. In the mobile application:
+
+   1. The user enters login and password
```
Comment (romanstrobl) [inline docs/Additional-Activation-OTP.md:62]: Missing period at the end of the sentence.

File: docs/Additional-Activation-OTP.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+In this common scenario, it's expected that the PowerAuth activation is not yet created so that the additional activation OTP can be used in the time of the activation creation as additional user authentication.
+
+### Regular activation with OTP
+
+1. User is authenticated in Master Front-End Application and initiates the activation creation process:
+
+   1. Master Front-End Application generates random activation OTP.
+   1. Master Front-End Application then asks PowerAuth server to create an activation, with using this OTP ([`initActivation`](https://github.com/wultra/powerauth-server/blob/develop/docs/SOAP-Service-Methods.md#method-initactivation) method, OTP validation set to ON_KEYS_EXCHANGE).
+   1. Master Front-End Application then displays QR code, containing an activation code.
+   1. At the same time, Master Front-End Application initiates the delivery of activation OTP. It's recommended to deliver such code via the dedicated out-of-band channel, for example, via SMS.
```
Comment (romanstrobl) [inline docs/Additional-Activation-OTP.md:32]: `via the dedicated out-of-band channel` -> `via a dedicated out-of-band channel`, we do not know yet what the exact channel is, `the` cannot be used yet

File: docs/Additional-Activation-OTP.md (hunk 1 above)
Comment (hvge) [reply docs/Additional-Activation-OTP.md:168]: I guess that `check the wether the` is also wrong (first "the" should be removed)

File: docs/Activation-Recovery.md [hunk 8]
```diff
@@ -46,14 +46,15 @@ The Activation via Recovery Code is by default **disabled** on PowerAuth Server.
 If application is activated via the Recovery Code and PUK, then it goes immediately to **ACTIVE** state (see [Activation States](Activation.md#activation-states)). There's no intermediate state when the user can confirm that activation in the additional digital channel (like internet banking). This fact limits this feature mostly to only limited application scenarios:
 
 1. Only for the application that doesn't provide active operations
-2. The application itself acts as a confirmation device (like our [Mobile Token](https://www.wultra.com/product/powerauth-mobile-token) does)
+2. The application itself acts as a confirmation device (like our [Mobile Token](https://www.wultra.com/mobile-token) does)
```
Comment (romanstrobl) [inline docs/Activation-Recovery.md:49]: Is this a lisp syntax? Not sure we need nested brackets.

File: docs/Additional-Activation-OTP.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+   1. The user scans QR code or retypes the activation code manually.
+   1. The user waits for OTP delivery via the out-of-band channel.
+   1. The user retypes OTP to the mobile application.
+   1. Mobile application then initializes the activation, using activation code and OTP. 
+
+1. Intermediate Server Application receives activation with activation code and OTP:
+ 
+   1. The activation code and OTP are verified by the PowerAuth server in the [`prepareActivation`](https://github.com/wultra/powerauth-server/blob/develop/docs/SOAP-Service-Methods.md#method-prepareactivation) method. 
+   1. If everything is right, the activation is set to the ACTIVE state. There's no need to wait for the confirmation.
```
Comment (romanstrobl) [inline docs/Additional-Activation-OTP.md:44]: I find the `everything is right` a little bit ambiguous. Perhaps we can say `if the method call succeeds` or something similar.

File: docs/Additional-Activation-OTP.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+   1. The user enters username and password.
+   1. The user waits for OTP delivery via the out-of-band channel.
+   1. The user retypes OTP.
+   1. Mobile application then initializes the custom activation with provided username, password, and OTP.
+
+1. Intermediate Server Application receives a custom activation request, with username, password, and OTP:
+
+   1. The username, password, and OTP is verified by the Intermediate Server Application.
+   1. If everything's right, then Intermediate Server Application creates activation by calling [`createActivation`](https://github.com/wultra/powerauth-server/blob/develop/docs/SOAP-Service-Methods.md#method-createactivation). The activation state can be set to ACTIVE.
```
Comment (romanstrobl) [inline docs/Additional-Activation-OTP.md:70]: Again `If everything's right` seems too ambiguous.

File: docs/Additional-Activation-OTP.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+1. In the mobile application: 
+
+   1. The user scans QR code or retypes the activation code manually.
+   1. Mobile application then initializes the activation, using activation code.
+
+1. Intermediate Server Application receives a regular activation request, with activation code:
+   
+   1. The activation code is verified by the PowerAuth server in the [`prepareActivation`](https://github.com/wultra/powerauth-server/blob/develop/docs/SOAP-Service-Methods.md#method-prepareactivation) method. 
+   1. If everything is right, the activation is set to the OTP_USED state.
```
Comment (romanstrobl) [inline docs/Additional-Activation-OTP.md:96]: dtto

**Pull request merged**